File: common/src/main/java/generations/gg/generations/core/generationscore/common/network/packets/S2COpenMailPacket.java
package generations.gg.generations.core.generationscore.common.network.packets;

import generations.gg.generations.core.generationscore.common.client.screen.mails.MailEditScreen;
import generations.gg.generations.core.generationscore.common.client.screen.mails.MailViewScreen;
import generations.gg.generations.core.generationscore.common.client.screen.mails.MailViewScreen.WrittenMailAccess;
import generations.gg.generations.core.generationscore.common.network.ClientNetworkPacketHandler;
import generations.gg.generations.core.generationscore.common.tags.GenerationsItemTags;
import net.minecraft.client.Minecraft;
import net.minecraft.network.FriendlyByteBuf;
import net.minecraft.resources.ResourceLocation;
import net.minecraft.world.InteractionHand;

import static generations.gg.generations.core.generationscore.common.GenerationsCore.id;

public record S2COpenMailPacket(InteractionHand hand) implements GenerationsNetworkPacket<S2COpenMailPacket> {
    public ResourceLocation getId() {
        return ID;
    }
    public void encode(FriendlyByteBuf buf) {
        buf.writeEnum(hand);
    }

    public static ResourceLocation ID = id("open_mail");
    public static S2COpenMailPacket decode(FriendlyByteBuf buf) {
        return new S2COpenMailPacket(buf.readEnum(InteractionHand.class));
    }

    public static class Handler implements ClientNetworkPacketHandler<S2COpenMailPacket> {
        public static final Handler INSTANCE = new Handler();
        public void handle(S2COpenMailPacket packet, Minecraft minecraft) {
            var itemStack = Minecraft.getInstance().player.getItemInHand(packet.hand());
            if (itemStack.is(GenerationsItemTags.CLOSED_POKEMAIL)) Minecraft.getInstance().setScreen(
                    new MailViewScreen(
                            new WrittenMailAccess(itemStack)));
            else if (itemStack.is(GenerationsItemTags.POKEMAIL)) Minecraft.getInstance().setScreen(
                    new MailEditScreen(
                            Minecraft.getInstance().player,
                        itemStack,
                        packet.hand)
            );
        }
    }
}

File: common/src/main/java/generations/gg/generations/core/generationscore/common/network/packets/S2COpenMailPacket.kt
package generations.gg.generations.core.generationscore.common.network.packets

import generations.gg.generations.core.generationscore.common.GenerationsCore
import net.minecraft.network.FriendlyByteBuf
import net.minecraft.resources.ResourceLocation
import net.minecraft.world.InteractionHand

@JvmRecord
data class S2COpenMailPacket(val hand: InteractionHand) : GenerationsNetworkPacket<S2COpenMailPacket> {
    override fun getId(): ResourceLocation {
        return ID
    }

    override fun encode(buf: FriendlyByteBuf) {
        buf.writeEnum(hand)
    }

    companion object {
        var ID = GenerationsCore.id("open_mail")
        fun decode(buf: FriendlyByteBuf): S2COpenMailPacket {
            return S2COpenMailPacket(buf.readEnum(InteractionHand::class.java))
        }
    }
}